File: 2015/go/day11/day11.go
package day11

import (
	"fmt"
	"io/ioutil"
)

//Day11 Corporate Policy
func Day11(inputFile string) error {
	input, err := ioutil.ReadFile(inputFile)
	if err != nil {
		return err
	}

	fmt.Println("Day 11: Corporate Policy")
	part1, err := part1(string(input))
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 1: %s\n", part1)

	part2, err := part2(part1)
	if err != nil {
		return err
	}
	fmt.Printf("\tPart 2: %s\n", part2)

	return nil
}

func checkPassword(password string) bool {
	oneAgo, twoAgo, threeAgo := '\u0000', '\u0000', '\u0000'
	pairs, triplet := 0, false
	for _, c := range password {
		switch c {
		case 'i', 'o', 'l':
			return false
		}
		if c-oneAgo == 1 && oneAgo-twoAgo == 1 {
			triplet = true
		}
		if c == oneAgo && (c != twoAgo || (twoAgo == threeAgo)) {
			pairs++
		}
		if triplet && pairs >= 2 {
			return true
		}
		threeAgo = twoAgo
		twoAgo = oneAgo
		oneAgo = c
	}
	return false
}

func incrementPassword(password string) string {
	for i := len(password) - 1; i >= 0; i-- {
		c := password[i]
		c++
		if c > 'z' {
			c = 'a'
		}
		newPassword := password[0:i] + string(c)
		if i < len(password)-1 {
			newPassword += password[i+1:]
		}
		if c != 'a' {
			return newPassword
		}
		password = newPassword
	}
	return password
}

func part1(input string) (string, error) {
	for !(checkPassword(input)) {
		input = incrementPassword(input)
	}
	return input, nil
}

func part2(input string) (string, error) {
	input = incrementPassword(input)
	for !(checkPassword(input)) {
		input = incrementPassword(input)
	}
	return input, nil
}

File: 2015/rust/src/day11.rs
use std::error::Error;
use std::fs::File;
use std::io::prelude::*;
use std::time::Instant;

pub fn day11(input_file: &str) -> Result<(), Box<dyn Error>> {
    println!("Day 11: Corporate Policy");

    let mut f = File::open(input_file)?;
    let mut input = String::new();
    f.read_to_string(&mut input)?;

    let start = Instant::now();
    let sol1 = part1(&input)?;
    println!("\tPart 1: {} ({:?})", sol1, start.elapsed());
    let second = Instant::now();
    println!("\tPart 2: {} ({:?})", part2(&sol1)?, second.elapsed());
    println!("\t\t Completed in {:?}", start.elapsed());
    Ok(())
}

fn increment(input: &mut Vec<u8>) {
    for c in input.into_iter().rev() {
        if char::from(*c) == 'z' {
            let mut b: [u8; 1] = [0];
            'a'.encode_utf8(&mut b[..]);
            *c = b[0];
        } else {
            *c += 1;
            return;
        }
    }
}

fn part1(input: &str) -> Result<String, Box<dyn Error>> {
    let mut input: Vec<u8> = Vec::from(input.as_bytes());
    loop {
        let (mut one_ago, mut two_ago, mut three_ago) = (0, 0, 0);
        let (mut pairs, mut triplet) = (0, false);
        for uc in input.iter() {
            let c = char::from(*uc);
            match c {
                'i' | 'o' | 'l' => break,
                _ => (),
            }
            if *uc as i32 - one_ago as i32 == 1 && one_ago as i32 - two_ago as i32 == 1 {
                triplet = true;
            }
            if *uc == one_ago && (*uc != two_ago || (two_ago == three_ago)) {
                pairs += 1;
            }
            if triplet && pairs >= 2 {
                return Ok(String::from_utf8(input)?);
            }
            three_ago = two_ago;
            two_ago = one_ago;
            one_ago = *uc;
        }
        increment(&mut input);
    }
}

fn part2(input: &str) -> Result<String, Box<dyn Error>> {
    let mut input: Vec<u8> = Vec::from(input.as_bytes());
    increment(&mut input);
    loop {
        let (mut one_ago, mut two_ago, mut three_ago) = (0, 0, 0);
        let (mut pairs, mut triplet) = (0, false);
        for uc in input.iter() {
            let c = char::from(*uc);
            match c {
                'i' | 'o' | 'l' => break,
                _ => (),
            }
            if *uc as i32 - one_ago as i32 == 1 && one_ago as i32 - two_ago as i32 == 1 {
                triplet = true;
            }
            if *uc == one_ago && (*uc != two_ago || (two_ago == three_ago)) {
                pairs += 1;
            }
            if triplet && pairs >= 2 {
                return Ok(String::from_utf8(input)?);
            }
            three_ago = two_ago;
            two_ago = one_ago;
            one_ago = *uc;
        }
        increment(&mut input);
    }
}
